src/ray/util/subreaper.cc

```diff
@@ -0,0 +1,177 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/subreaper.h"
+
+#include <string.h>
+
+#include <vector>
+
+#ifdef __linux__
+#include <sys/prctl.h>
+#include <sys/wait.h>
+#endif
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/util/logging.h"
+#include "ray/util/macros.h"
+#include "ray/util/process.h"
+
+namespace ray {
+// Platform-specific implementation of subreaper code.
+
+#ifdef __linux__
+// Linux implementation.
+
+namespace {
+
+using SignalHandlerFn = void (*)(const boost::system::error_code &, int);
+// Register a signal handler for the given signal.
+// The handler will be called with the signal_set and the error code.
+// After the handler is called, the signal will be re-registered.
+// The callback keeps a reference of the shared ptr to make sure it's not destroyed.
+void RegisterSignalHandlerLoop(std::shared_ptr<boost::asio::signal_set> signals,
+                               SignalHandlerFn handler) {
+  signals->async_wait(
+      [signals, handler](const boost::system::error_code &error, int signal_number) {
+        handler(error, signal_number);
+        RegisterSignalHandlerLoop(signals, handler);
+      });
+}
+
+void SigchldHandlerReapZombieAndRemoveKnownChildren(
+    const boost::system::error_code &error, int signal_number) {
+  if (error) {
+    RAY_LOG(WARNING) << "Error in SIGCHLD handler: " << error.message();
+  }
+  int status;
+  pid_t pid;
+  // Reaps any children that have exited. WNOHANG makes waitpid non-blocking and returns
+  // 0 if there's no zombie children.
+  while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
```

Comment:
this should be non-blocking right? (since proc should have already dead)

Reply:
Non-blocking, yes. It's controlled by the WNOHANG. If there's no child dead already, it returns -1